From d9001fe74288afe1ae90b4ca7d923e740ad81616 Mon Sep 17 00:00:00 2001
From: Hanjun Guo <guohanjun@huawei.com>
Date: Wed, 29 Sep 2021 16:02:10 +0800
Subject: [PATCH] arm64: phytium: using MIDR_PHYTIUM_FT2000PLUS instead of
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: d9001fe74288afe1ae90b4ca7d923e740ad81616

 ARM_CPU_IMP_PHYTIUM

hulk inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I41AUQ
CVE: NA

-------------------------------------

Update the code to using MIDR_PHYTIUM_FT2000PLUS, instead of
ARM_CPU_IMP_PHYTIUM, which will distinguish FTC662 and FTC663.

Signed-off-by: Hanjun Guo <guohanjun@huawei.com>
Cc: Guo Hui <guohui@uniontech.com>
Cc: Cheng Jian <cj.chengjian@huawei.com>
Cc: Zhen Lei <thunder.leizhen@huawei.com>
Cc: Xiuqi Xie <xiexiuqi@huawei.com>
Reviewed-by: Xie XiuQi <xiexiuqi@huawei.com>
Signed-off-by: Cheng Jian <cj.chengjian@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/arm64/kernel/topology.c | 6 ++++--
 drivers/usb/host/xhci-pci.c  | 4 +++-
 2 files changed, 7 insertions(+), 3 deletions(-)

diff --git a/arch/arm64/kernel/topology.c b/arch/arm64/kernel/topology.c
index e5c9ac4840c6..7a487976ce8b 100644
--- a/arch/arm64/kernel/topology.c
+++ b/arch/arm64/kernel/topology.c
@@ -53,9 +53,11 @@ void store_cpu_topology(unsigned int cpuid)
 	cpuid_topo->thread_id  = -1;
 	cpuid_topo->core_id    = cpuid;
 	cpuid_topo->package_id = cpu_to_node(cpuid);
-	if (read_cpuid_implementor() == ARM_CPU_IMP_PHYTIUM) {
+
+	/* Some PHYTIUM FT2000PLUS platform firmware has no PPTT table */
+	if ((read_cpuid_id() & MIDR_CPU_MODEL_MASK) == MIDR_PHYTIUM_FT2000PLUS
+		&& cpu_to_node(cpuid) == NUMA_NO_NODE) {
 		cpuid_topo->thread_id  = 0;
-		cpuid_topo->core_id  = cpuid;
 		cpuid_topo->package_id = 0;
 	}
 
diff --git a/drivers/usb/host/xhci-pci.c b/drivers/usb/host/xhci-pci.c
index 3fe73d2f39e9..8b15c2644de6 100644
--- a/drivers/usb/host/xhci-pci.c
+++ b/drivers/usb/host/xhci-pci.c
@@ -375,8 +375,10 @@ static int xhci_pci_setup(struct usb_hcd *hcd)
 #include <asm/cputype.h>
 static void phytium_xhci_pci_workaround(struct pci_dev *dev)
 {
+	u32 midr = read_cpuid_id();
+
 	/* Firmware bug, DMA mask is not reported by the firmware */
-	if (read_cpuid_implementor() == ARM_CPU_IMP_PHYTIUM)
+	if ((midr & MIDR_CPU_MODEL_MASK) == MIDR_PHYTIUM_FT2000PLUS)
 		dma_set_mask(&dev->dev, DMA_BIT_MASK(64));
 }
 #else
-- 
2.26.2

